File: product-brain/src/product_brain/backfill/summarize.py
from __future__ import annotations

from typing import Callable, Optional

from ..config import Config
from ..llm import get_provider


def llm_call_factory(config: Config, model: Optional[str] = None) -> Callable[[str], str]:
    """Return a function that takes a prompt and returns text.

    Provider is chosen via config.llm.provider (anthropic | openai |
    azure_openai | openai_compatible). All product-brain LLM usage flows
    through this factory.
    """
    provider = get_provider(config)
    chosen_model = model or config.llm.model_summarize

    def call(prompt: str, max_tokens: int = 2000) -> str:
        return provider.call(prompt, chosen_model, max_tokens=max_tokens)

    return call

File: product-brain-ts/src/backfill/summarize.ts
// Port target: ../product-brain/src/product_brain/backfill/summarize.py
import type { Config } from "../config.js";
import { getProvider } from "../llm/index.js";

export type LLMCall = (prompt: string, maxTokens?: number) => Promise<string>;

export function llmCallFactory(config: Config, model?: string): LLMCall {
  const provider = getProvider(config);
  const chosenModel = model ?? config.llm.model_summarize;
  return async (prompt, maxTokens = 2000) => provider.call(prompt, chosenModel, maxTokens);
}
